File: leetcode-js/155.js

var MinStack = function () {
    this.stack = [];
    this.minVal = [];
};

/** 
 * @param {number} val
 * @return {void}
 */
MinStack.prototype.push = function (val) {
    this.stack.push(val);
    if (this.minVal.length > 0) {
        this.minVal.push(Math.min(this.getMin(), val));
    }
    else this.minVal.push(val);
};

/**
 * @return {void}
 */
MinStack.prototype.pop = function () {
    this.stack.pop();
    this.minVal.pop();
};

/**
 * @return {number}
 */
MinStack.prototype.top = function () {
    return this.stack[this.stack.length - 1];
};

/**
 * @return {number}
 */
MinStack.prototype.getMin = function () {
    return this.minVal[this.minVal.length - 1];
};

/** 
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(val)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */

minStack = new MinStack();
minStack.push(0);
minStack.push(1);
minStack.push(0);
console.log(minStack.getMin()); // return -3
minStack.pop();
console.log(minStack.top());    // return 0
console.log(minStack.getMin()); // return -2

File: leetcode-go/155.go
package main

import "math"

type MinStack struct {
	Values        []int
	MinValueTrack []int
}

// func Constructor() MinStack {
// 	return MinStack{
// 		Values:        []int{},
// 		MinValueTrack: []int{},
// 	}
// }

func (this *MinStack) Push(val int) {
	this.Values = append(this.Values, val)
	minValue := val
	if len(this.MinValueTrack) > 0 {
		minValue = int(math.Min(float64(this.MinValueTrack[len(this.MinValueTrack)-1]), float64(val)))
	}
	this.MinValueTrack = append(this.MinValueTrack, minValue)
}

func (this *MinStack) Pop() {
	if len(this.Values) > 0 {
		this.Values = this.Values[:len(this.Values)-1]
		this.MinValueTrack = this.MinValueTrack[:len(this.MinValueTrack)-1]
	}
}

func (this *MinStack) Top() int {
	return this.Values[len(this.Values)-1]
}

func (this *MinStack) GetMin() int {
	return this.MinValueTrack[len(this.MinValueTrack)-1]
}

/**
 * Your MinStack object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(val);
 * obj.Pop();
 * param_3 := obj.Top();
 * param_4 := obj.GetMin();
 */
